src/main/java/org/truffleruby/core/proc/ProcNodes.java

```diff
@@ -274,6 +276,37 @@ protected RubyProc createSameArityProc(RubyProc userProc, RubyProc block) {
         }
     }
 
+    @Primitive(name = "proc_specify_arity", lowerFixnum = 1)
+    public abstract static class ProcSpecifyArityNode extends PrimitiveArrayArgumentsNode {
+        @Specialization
+        protected RubyProc specifyArityProc(RubyProc block, int argc) {
+            Arity oldArity = block.sharedMethodInfo.getArity();
+            final Arity newArity;
+            if (argc <= -1) {
+                newArity = new Arity(0, oldArity.getOptional(), true);
+            } else {
+                newArity = new Arity(argc, oldArity.getOptional(), false);
```

Comment:
Let's pass `0` for optional, instead of `oldArity.getOptional()`, also above.
From a single arity number we cannot know if there are optional arguments or a rest argument, but here we want a rest argument.
```ruby
[8] pry(main)> -> a,b=1,c=2{}.arity
=> -2
[9] pry(main)> -> a,*r{}.arity
=> -2
```